lib/fluent/plugin/elasticsearch_error_handler.rb

```diff
@@ -3,23 +3,33 @@
 class Fluent::ElasticsearchErrorHandler
   include Fluent::ElasticsearchConstants
 
-  attr_accessor :records, :bulk_message_count
   class ElasticsearchVersionMismatch < StandardError; end
   class ElasticsearchError < StandardError; end
 
-  def initialize(plugin, records = 0, bulk_message_count = 0)
+  def initialize(plugin)
     @plugin = plugin
-    @records = records
-    @bulk_message_count = bulk_message_count
   end
 
-  def handle_error(response, tag, records)
-    if records.length != response['items'].length
+  def handle_error(response, tag, chunk, bulk_message_count)
+    if bulk_message_count != response['items'].length
       raise ElasticsearchError, "The number of records submitted do not match the number returned. Unable to process bulk response"
     end
     retry_records = []
     stats = Hash.new(0)
-    response['items'].each_with_index do |item, index|
+    meta = {}
+    header = {}
+    chunk.msgpack_each do |rawrecord, time|
+      bulk_message = ''
+      next unless rawrecord.is_a? Hash
+      begin
+        # we need a deep copy for process_message to alter
+        processrecord = Marshal.load(Marshal.dump(rawrecord))
```

Comment:
Does `rawrecord.dup` not perform the same function without serializing and deserializing?

Reply:
>  Is that because we spin through the chunk and process_message again after having done that already in write_obj

Yes.  And in the tests I added, you can see that it will correctly associate messages from the chunk with the responses from Elasticsearch.

> I'm not sure we have any memory savings given we are duplicating records anyway

In the _error_ case, no, we are not saving any memory.  However, in the _regular_ case, we are saving memory - we are not adding any code or any extra memory in the _regular_, non-error case.  That is, this patch makes it so that we do not have any effect on the regular, non-error case, which is what @portante was concerned about.  Then, in the error case, yes, we use additional memory - we pay the price in the error case, to avoid having to pay the price in the non-error case.

> Does rawrecord.dup not perform the same function without serializing and deserializing?

No, it does a shallow copy, and we need a deep copy here.

Reply:
BTW, I've also tested this code with logging, and it passes the bulk rejection test.